### PR TITLE
Ch 506: Allow asynchronous visitor context processing

### PR DESCRIPTION
**Pull request by eshtadc**

### CODE REVIEW

File: js/personalize.js [hunk 1]
```diff
@@ -222,56 +288,76 @@
...
+   *   The plugin name to use to retrieve visitor context.
+   * @param context
+   *   The context object to retrieve values for in the current visitor context.
+   * @returns {Promise}
+   *   An asynchronous JS promise.
+   */
+  function getVisitorContext(plugin, context) {
+    var visitor_context = Drupal.personalize.visitor_context;
+    if (visitor_context.hasOwnProperty(plugin) && typeof visitor_context[plugin].getContext === 'function') {
+      return context_values = visitor_context[plugin].getContext(context);
```
Comment (katbailey) [inline js/personalize.js]: I think this assignment to the context_values variable is leftover from the old code..?

File: js/personalize.js [hunk 2]
```diff
@@ -222,56 +288,76 @@
...
 
+  /**
+   * Gets the values for a particular visitor context.
+   *
+   * @param plugin
+   *   The plugin name to use to retrieve visitor context.
+   * @param context
+   *   The context object to retrieve values for in the current visitor context.
+   * @returns {Promise}
+   *   An asynchronous JS promise.
```
Comment (katbailey) [inline js/personalize.js]: Could also be a hash of context values or null, right?

File: js/personalize.js [hunk 3]
```diff
@@ -677,45 +753,147 @@
...
-    agents[agent_name].decisionPoints[decision_point] =
-       agents[agent_name].decisionPoints[decision_point] || { choices: {}, callbacks: {}, fallbacks: {}};
+    agentData.decisionPoints[decision_point] = { choices: {}, callbacks: {}, fallbacks: {}};
+
+    agentData.decisionPoints[decision_point].choices[decision_name] = choices;
+    agentData.decisionPoints[decision_point].fallbacks[decision_name] = fallbackIndex;
+    addDecisionCallback(executor, agent_name, decision_point, $option_set, osid);
+
+    // Check to see if this decision is already in storage.
+    decisions = readDecisionsFromStorage(agent_name, decision_point);
```
Comment (katbailey) [inline js/personalize.js]: This needs a var declaration (it got moved out of the function that originally had the var declaration).

File: js/personalize.js [hunk 4]
```diff
@@ -483,7 +574,7 @@
   }
 
   function writeDecisionsToStorage(agent_name, point, decisions) {
-    if (!sessionID || !Drupal.settings.personalize.agent_map[agent_name].cache_decisions) {
+    if (!Drupal.personalize.sessionId || !Drupal.settings.personalize.agent_map[agent_name].cache_decisions) {
```
Comment (katbailey) [inline js/personalize.js]: Unrelated to the changes here, but do we still need a session ID to store decisions? Currently this means that if you're using personalize_target module the decisions don't get cached. This is probably usually the desired behavior, but it seems to be that it is by accident (it should just rely on the cache_decisions setting)

Comment (eshtadc) [inline js/personalize.js]: No we don't actually - I'll take it out here .

**Comment by eshtadc** [comment]
merged.

**Pull request closed without merging**